daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -815,12 +816,16 @@ private[lf] object SBuiltin {
     }
   }
 
+  // SBStructCon sorts the field after evaluation of its arguments to preserve
+  // evaluation order of unordered fields.
   /** $tcon[fields] :: a -> b -> ... -> Struct */
-  final case class SBStructCon(fields: ImmArray[Name])
-      extends SBuiltinPure(fields.length)
-      with SomeArrayEquals {
+  final case class SBStructCon(inputFieldsOrder: Struct[Int])
+      extends SBuiltinPure(inputFieldsOrder.size) {
+    private[this] val fieldNames = inputFieldsOrder.mapValues(_ => ())
     override private[speedy] final def executePure(args: util.ArrayList[SValue]): SValue = {
-      SStruct(fields, args)
+      val sortedFields = new util.ArrayList[SValue](inputFieldsOrder.size)
+      inputFieldsOrder.values.foreach(i => sortedFields.add(args.get(i)))
+      SStruct(fieldNames, sortedFields)
```

Comment:
I like how we pushed the sorting to compile time and keep it cheap at runtime.

I'm wondering if there's a way to avoid allocating a second array and reuse the `args` array instead? I couldn't come up quickly with a simple way myself though.

Reply:
The simplest I could find is as follow. Still not sure it is worthwhile.

```scala
  // SBStructCon sorts the field after evaluation of its arguments to preserve
  // evaluation order of unordered fields.
  /** $tcon[fields] :: a -> b -> ... -> Struct */
  final case class SBStructCon(inputFieldsOrder: Struct[Int])
      extends SBuiltinPure(inputFieldsOrder.size) {

    override private[speedy] final def executePure(args: util.ArrayList[SValue]): SValue = {
      order(args)
      SStruct(fieldNames, args)
    }

    private[this] val fieldNames = inputFieldsOrder.mapValues(_ => ())

    private[this] val permutations: List[(Int, List[Int])] = {

      val mapping = Array.ofDim[Int](inputFieldsOrder.size)
      mapping.indices.foreach { i =>
        mapping(inputFieldsOrder.toImmArray(i)._2) = i
      }

      @tailrec
      def loop1(toVisit: Set[Int], acc: List[(Int, List[Int])]): List[(Int, List[Int])] =
        if (toVisit.isEmpty)
          acc
        else {
          val next = toVisit.head
          val (nextToVisit, perm @ (_, others)) = loop2(toVisit - next, next, List.empty)
          val nextAcc = if (others.isEmpty) acc else perm :: acc
          loop1(nextToVisit, nextAcc)
        }

      @tailrec
      def loop2(toVisit: Set[Int], curr: Int, previous: List[Int]): (Set[Int], (Int, List[Int])) = {
        val next = mapping(curr)
        if (toVisit(next))
          loop2(toVisit - next, next, curr :: previous)
        else
          (toVisit, (curr, previous))
      }

      loop1(mapping.toSet, List.empty)
    }

    @inline
    private[this] def order[X](a: util.ArrayList[X]): Unit =
      permutations.foreach {
        case (first, others) =>
          var i = first
          val tmp = a.get(i)
          others.foreach { j =>
            a.set(i, a.get(j))
            i = j
          }
          a.set(i, tmp)
      }
  }
```

Reply:
This seems to be a lot of complexity for a gain we can't properly quantify yet. Thus, I'd say it's not worthwhile for now.